test/ao/sparsity/test_structured_sparsifier.py

```diff
@@ -916,3 +917,123 @@ def test_prune_lstm_layernorm_linear_single_layer(self):
+
+    def _test_update_mask_on_single_layer(self, expected_conv1, device):
+        """Test that pruning is conducted based on the pair-wise distance measurement instead of absolute norm value"""
+        # test pruning with one layer of conv2d
+        model = TestFPGMPruner.SimpleConvFPGM().to(device)
+        x = torch.ones((1, 1, 32, 32), device=device)
+        pruner = FPGMPruner(0.3)
+        config = [{"tensor_fqn": "conv2d1.weight"}]
+        pruner.prepare(model, config)
+        pruner.enable_mask_update = True
+        pruner.step()
+        assert pruner.groups[0]["module"].parametrizations.weight[0].mask[-1].item() is not False,\
+            "do not prune the least-norm filter"
+
+        # fusion step
+        pruned_model = pruner.prune()
+
+        pruned_y = pruned_model(x)
+        # assert shapes
+        assert pruned_y.shape == (1, 4, 32, 32)
+        assert pruned_model.conv2d1.weight.shape == expected_conv1.shape
+        assert pruned_model.conv2d2.weight.shape == (4, 2, 3, 3), "conv2d2 should have input channel pruned"
+        # assert value
+        assert torch.isclose(pruned_model.conv2d1.weight, expected_conv1, rtol=1e-05, atol=1e-07).all()
+
+    def _test_update_mask_on_multiple_layer(self, expected_conv1, expected_conv2, device):
+        # the second setting
+        model = TestFPGMPruner.SimpleConvFPGM().to(device)
+        x = torch.ones((1, 1, 32, 32), device=device)
+        pruner = FPGMPruner(0.3)
+        config = [
+            {"tensor_fqn": "conv2d1.weight"},
+            {"tensor_fqn": "conv2d2.weight", "sparsity_level": 0.5}
+        ]
+        pruner.prepare(model, config)
+        pruner.enable_mask_update = True
+        pruner.step()
+        # Get the masks for the two least-norm filters
+        mask1 = pruner.groups[0]['module'].parametrizations.weight[0].mask[-1]
+        mask2 = pruner.groups[0]['module'].parametrizations.weight[0].mask[-2]
+        # Check if either of the least-norm filters is not pruned
+        assert mask1.item() is not False or mask2.item() is not False, "Do not prune all least-norm filters"
+
+        # fusion step
+        pruned_model = pruner.prune()
+        pruned_y = pruned_model(x)
+        # assert shapes
+        assert pruned_y.shape == (1, 2, 32, 32)
+        assert pruned_model.conv2d1.weight.shape == expected_conv1.shape
+        assert pruned_model.conv2d2.weight.shape == expected_conv2.shape
+        # assert values
+        assert torch.isclose(pruned_model.conv2d1.weight, expected_conv1, rtol=1e-05, atol=1e-07).all()
+        assert torch.isclose(pruned_model.conv2d2.weight, expected_conv2, rtol=1e-05, atol=1e-07).all()
+
+    def test_update_mask(self):
+        weights = torch.tensor([3.0, 0.1])
+        expected_conv1 = torch.ones((2, 1, 3, 3)) * weights[:, None, None, None]
+
+        weights = torch.tensor([7.0, 0.4])
+        expected_conv2 = torch.ones((2, 2, 3, 3)) * weights[:, None, None, None]
+
+        for device in DEVICES:
```

Comment:
@ArmandXiao I think you just need to move expected_conv1 and expected_conv2 to the right device.

Reply:
Hi @jcaip , thank you a lot for the information.

In the [latest commit](https://github.com/pytorch/pytorch/pull/95689/commits/1c98cbece24122e3a13c44397f7586aa81ec2e6a), I have move `expected_conv1` and `expected_conv2` to the correct device before the test cases.